tcmur_cmd_handler.c

```diff
@@ -330,12 +392,51 @@ static int handle_write_same(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
 	write_same->iov_cnt = 1;
 	cmd->cmdstate = write_same;
 
-	cmd->done = handle_write_same_cbk;
+	cmd->done = handle_writesame_cbk;
 
 	tcmu_dev_dbg(dev, "First lba: %llu, write lbas: %llu\n",
 		     start_lba, write_lbas);
 
-	return async_handle_cmd(dev, cmd, write_same_work_fn);
+	return async_handle_cmd(dev, cmd, writesame_work_fn);
+}
+
+static int tcmur_writesame_work_fn(struct tcmu_device *dev,
+				 struct tcmulib_cmd *cmd)
+{
+	struct tcmur_handler *rhandler = tcmu_get_runner_handler(dev);
+	tcmur_writesame_fn_t write_same_fn = cmd->cmdstate;
+	uint32_t block_size = tcmu_get_dev_block_size(dev);
+	uint8_t *cdb = cmd->cdb;
+	uint64_t off = block_size * tcmu_get_lba(cdb);
+	uint32_t len = block_size * tcmu_get_xfer_length(cdb);
+	bool unmap = false;
+
+	if (rhandler->unmap && (cmd->cdb[1] & 0x08)) {
+		tcmu_dev_dbg(dev, "Do UNMAP in WRITE_SAME cmd!\n");
+		unmap = true;
```

Comment:
It looks like you can drop the unmap code in rbd_aio_writesame, move the setting of the done function to handle_generic_cbk before the "if" above and then call rhandler->unmap here.